presto-raptorx/src/main/java/com/facebook/presto/raptorx/TransactionManager.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptorx;
+
+import com.facebook.presto.raptorx.metadata.Metadata;
+import com.facebook.presto.raptorx.transaction.Transaction;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+
+import javax.inject.Inject;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.facebook.presto.raptorx.RaptorErrorCode.RAPTOR_INTERNAL_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class TransactionManager
+{
+    private final Metadata metadata;
+
+    private final Map<ConnectorTransactionHandle, Transaction> transactions = new ConcurrentHashMap<>();
+
+    @Inject
+    public TransactionManager(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    public ConnectorTransactionHandle create()
+    {
+        long transactionId = metadata.nextTransactionId();
+        long commitId = metadata.getCurrentCommitId();
```

Comment:
Let me describe a potential data race here. Let's verify together whether it's a real issue.
The race happens when some logics jump in after line 44 and before line 48, the time series is:
thread2      called beginTransaction, get CurrentCommitId == 1
thread1      metadata.getCurrentCommitId();  get CurrentCommitId==1
thread2     execute, finish commit, update CurrentCommitId=2, exit fully
CommitCleaner threadxx  called oldestActiveCommitId,  since transactions is empty, CurrentCommitId==2, it gets 
   oldestActiveCommitId == 2
thread1    called transactions.put(handle, transaction);

Then this enters a unsafe state, since oldestActiveCommitId is actually 1,   but CommitCleaner get 2.

Reply:
Nice find! You are correct.